### PR TITLE
MABL-4397: add proxy support

### PR DESCRIPTION
**Pull request by bertold**
Add proxy support with optional proxy credentials stored inside the Jenkins credentials plugin as username/password secret
Add health check to test connectivity into the validate form method
Update tests
Update components

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [X] TODO: Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/com/mabl/integration/jenkins/validation/MablStepBuilderValidator.java [hunk 1]
```diff
@@ -1,6 +1,14 @@
 package com.mabl.integration.jenkins.validation;
 
+import com.mabl.integration.jenkins.MablRestApiClient;
+import com.mabl.integration.jenkins.MablRestApiClientImpl;
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/validation/MablStepBuilderValidator.java:4]: This appears unused.

**Review by twistedpair — COMMENTED** [review]
Is there a reason we cannot use the Jenkins Plugins Proxy API like [peer plugins do](https://github.com/jenkinsci/fortify-on-demand-uploader-plugin/blob/5d096f48e7e49855fcbf6d252968415e82e4e6db/src/main/java/org/jenkinsci/plugins/fodupload/FodApiConnection.java#L54)? Seems like we're writing our own plugin proxy management tooling here, while it appears to already exist in Jenkins.

image

image

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 2]
```diff
@@ -311,7 +312,7 @@ private RequestConfig getDefaultRequestConfig() {
                 .setConnectTimeout(REQUEST_TIMEOUT_MILLISECONDS)
                 .setConnectionRequestTimeout(REQUEST_TIMEOUT_MILLISECONDS)
                 .setSocketTimeout(REQUEST_TIMEOUT_MILLISECONDS)
-                .setProxy(proxy)
+                .setProxy(proxy != null ? new HttpHost(proxy.name, proxy.port) : null)
                 .setProxyPreferredAuthSchemes(Collections.singletonList(AuthSchemes.BASIC))
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:316]: Curious, are we assuming all proxies will use base auth? Perhaps that's the only format supported by Jenkins plugins?

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 3]
```diff
@@ -223,6 +259,23 @@ public String getAppBaseUrl() {
...
+    @Override
+    public void checkConnection() throws IOException {
+        final String url = restApiBaseUrl + HEALTH_ENDPOINT;
+        final HttpResponse response = httpClient.execute(buildGetRequest(url));
+        final int statusCode = response.getStatusLine().getStatusCode();
+        switch (statusCode) {
+            case 200:
+                break;
+            case 401:
+                throw new IOException("invalid API key provided or the proxy connection requires authentication");
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:271]: Minor: can we capitalize these?
```suggestion
                throw new IOException("Invalid API key provided or the proxy connection requires authentication");
```

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 4]
```diff
@@ -223,6 +259,23 @@ public String getAppBaseUrl() {
...
+            case 403:
+                throw new IOException("invalid API key type. You must use a CI/CD Integration key type");
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:273]:
```suggestion
                throw new IOException("Invalid API key type. You must use a CI/CD Integration key type");
```

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 5]
```diff
@@ -153,9 +167,31 @@ private CredentialsProvider getApiCredentialsProvider(
 
         provider.setCredentials(AuthScope.ANY, creds);
 
+        // Set proxy credentials if provided
+        if (proxy != null && !StringUtils.isBlank(proxy.getUserName())) {
+            final Credentials c = new UsernamePasswordCredentials(
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:172]: Minor: we try not to use single letter var names.

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 6]
```diff
@@ -153,9 +167,31 @@ private CredentialsProvider getApiCredentialsProvider(
...
     }
 
+    private CredentialsProvider getProxyCredentialsProvider(
+            final StandardUsernamePasswordCredentials credentials
+    ) {
+        final CredentialsProvider provider = new BasicCredentialsProvider();
+        final UsernamePasswordCredentials creds =
+                new UsernamePasswordCredentials(REST_API_USERNAME_PLACEHOLDER, restApiKey.getPlainText());
+
+        provider.setCredentials(AuthScope.ANY, creds);
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:189]: How's this work. Do more specific auth scopes (like proxy creds) override less specific cred scopes?

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 7]
```diff
@@ -223,6 +259,23 @@ public String getAppBaseUrl() {
         return appBaseUrl;
     }
 
+    @Override
+    public void checkConnection() throws IOException {
+        final String url = restApiBaseUrl + HEALTH_ENDPOINT;
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:264]: I don't think we should use this endpoint. It's meant for internal checks and is quite expensive (makes calls to internal resources to determine overall API health). If we have a need to determine the API is up, I think we should choose a different endpoint. 

A liveness check endpoint would be a better candidate (e.g. is internal NOOP, `/health/live`).

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 8]
```diff
@@ -153,9 +167,31 @@ private CredentialsProvider getApiCredentialsProvider(
 
         provider.setCredentials(AuthScope.ANY, creds);
 
+        // Set proxy credentials if provided
+        if (proxy != null && !StringUtils.isBlank(proxy.getUserName())) {
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:171]: Cloud blank usernames be considered value for someone's proxy configuration? (e.g. password only)?

File: src/main/resources/com/mabl/integration/jenkins/MablStepBuilder/help-proxyCredentials.html [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<div>
```
Comment (twistedpair) [inline src/main/resources/com/mabl/integration/jenkins/MablStepBuilder/help-proxyCredentials.html:1]: Do we still need these help HTML refs, if we're using the internal Jenkins proxy configs?

File: src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java [hunk 10]
```diff
@@ -198,6 +199,106 @@ public void createDeploymentRequestWithBranch() throws IOException {
         assertSuccessfulCreateDeploymentRequest(fakeRestApiKeyId, environmentId, applicationId, null, branch);
     }
 
+    @Test
+    public void testCheckConnection_Ok() throws IOException {
```
Comment (twistedpair) [inline src/test/java/com/mabl/integration/jenkins/MablRestApiClientTest.java:203]: Nice tests.

**Review by twistedpair — COMMENTED** [review]
I :heart:  it. Much simpler :smile:

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 11]
```diff
@@ -153,9 +167,31 @@ private CredentialsProvider getApiCredentialsProvider(
...
+                    proxy.getPassword()
+            );
+            provider.setCredentials(new AuthScope(new HttpHost(proxy.name, proxy.port)), c);
+        }
+
         return provider;
     }
 
+    private CredentialsProvider getProxyCredentialsProvider(
+            final StandardUsernamePasswordCredentials credentials
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:183]: Minor: unused var.

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java (hunk 8 above)
Comment (bertold) [reply src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:171]: I have not seen any case like this, but I will update the code. The UsernamePasswordCredentials class does not allow null values for the username, but I can pass in an empty string.

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java (hunk 6 above)
Comment (bertold) [reply src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:189]: Unused method. I left this in by mistake.

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java (hunk 2 above)
Comment (bertold) [reply src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:316]: Yes - that's the only way I found it in existing plugins. Some don't even support proxy authentication.

File: src/main/resources/com/mabl/integration/jenkins/MablStepBuilder/help-proxyCredentials.html (hunk 9 above)
Comment (bertold) [reply src/main/resources/com/mabl/integration/jenkins/MablStepBuilder/help-proxyCredentials.html:1]: Yes as I added a call to the health/live endpoint as part of the form validation (which was not there before).

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 12]
```diff
@@ -153,6 +164,15 @@ private CredentialsProvider getApiCredentialsProvider(
 
         provider.setCredentials(AuthScope.ANY, creds);
 
+        // Set proxy credentials if provided
+        if (proxy != null && (!StringUtils.isBlank(proxy.getUserName()) || !StringUtils.isBlank(proxy.getPassword()))) {
+            final Credentials credentials = new UsernamePasswordCredentials(
+                    proxy.getUserName() == null ? "" : proxy.getUserName(),
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:170]: Is this null handler needed for passwords too?

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java [hunk 13]
```diff
@@ -223,6 +243,23 @@ public String getAppBaseUrl() {
...
+        final String url = restApiBaseUrl + HEALTH_LIVE_ENDPOINT;
+        final HttpResponse response = httpClient.execute(buildGetRequest(url));
+        final int statusCode = response.getStatusLine().getStatusCode();
+        switch (statusCode) {
+            case 200:
+                break;
+            case 401:
+                throw new IOException("Invalid API key provided or the proxy connection requires authentication");
+            case 403:
+                throw new IOException("Invalid API key type. You must use a CI/CD Integration key type");
```
Comment (twistedpair) [inline src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:257]:
```suggestion
                throw new IOException("Invalid API key type. You must use a 'CI/CD Integration' key type");
```

Maybe we add quotes to help them realize this is the explicit key type name.

**Review by twistedpair — APPROVED** [review]
Great work, @bertold ! :tada:

File: src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java (hunk 12 above)
Comment (bertold) [reply src/main/java/com/mabl/integration/jenkins/MablRestApiClientImpl.java:170]: I looked at the implementation and it was not necessary.

**Pull request merged**